src/ucp/api/ucp_def.h

```diff
@@ -771,5 +771,60 @@ typedef struct ucp_ep_params {
  */
 #define UCP_ENTITY_NAME_MAX 32
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief A transport name and device name pair used by this endpoint. The
+ * caller is responsible for the allocation and deallocation of an array of
+ * these structures large enough to contain the desired number of transport
+ * and device name pairs.
+ */
+typedef struct {
+    /**
+     * The name of a transport layer used by this endpoint. This string is
+     * valid until the endpoint is shut down.
+     */
+    const char *transport_name;
+
+    /**
+     * The name of the device used with this transport by this endpoint.
+     * This string is valid until the endpoint is shut down.
+     */
+    const char *device_name;
+
+} ucp_transport_entry_t;
+
+
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief Structure containing an array of transport layers and device names
+ * used by an endpoint.
+ *
+ * The caller is responsible for allocation and deallocation of 
+ * this structure.
+ */
+typedef struct {
+    /**
+      * Pointer to array of transport/device name pairs used by this endpoint.
+      * The caller is responsible for the allocation and deallocation of this
+      * array.
+      */
+    ucp_transport_entry_t *entries;
+
+    /**
+     * Number of transport/device name pairs. The caller must set this to
+     * the maximum number of these pairs the structure can contain. On return,
+     * this is set to the actual number of transport and device name pairs used
+     * by the endpoint.
+     */
+    unsigned              num_entries;
+
+    /**
+     * Size of a single ucp_transport_entry_t object. The caller sets this to
+     * the size of the ucp_transport_entry_t they are using. UCP code must not
+     * set any fields in ucp_transport_entry_y beyond this size.
+     */
+    size_t                entry_size;
```

Comment:
@yosefe Do you think this provides better backward compatibility support compared to the bit-field in `ucp_transport_entry_t` ? If so, why ? thanks

Reply:
entry size allows knowing the size of each entry from caller perspective so UCX will try to fill only valid fields.
field_mask is not usable here because need to access elements by array index

Reply:
@yosefe I agree that you need size to calculate index and etc. I don't this size alone is enough to know what fields is okay to fill in.

Reply:
assuming we always add fields to the end of the struct, it should be enough IMO. 
Seems rdma-core is doing the same here https://github.com/linux-rdma/rdma-core/blob/486ecb3f12ab17e4b7970a6d5444cd165cec6ee4/libibverbs/cmd_device.c#L484

Reply:
If we go the route that we only append new fields to the end of the struct we need explicitly document how the struct can be safely extended. We also need add a code example how to use the API. This is the first time we use the sizeof trick and it is not as obvious.

Reply:
@drwootton Follow up on the conversation today. If you grep through UCP.H for a `@code` you will see few example codes that demonstrate a bit more complex features. I think it will be super useful to have small example code demonstrate how to use this.